crates/sui/src/client_commands.rs

```diff
@@ -1002,13 +1003,28 @@ impl SuiClientCommands {
                         gas_budget,
                     )
                     .await?;
-                serialize_or_execute!(
+                let result = serialize_or_execute!(
                     data,
                     serialize_unsigned_transaction,
                     serialize_signed_transaction,
                     context,
                     Upgrade
-                )
+                );
+                if let SuiClientCommandResult::Upgrade(ref response) = result {
+                    let build_config = resolve_lock_file_path(build_config, Some(package_path))?;
+                    if let Err(e) = sui_package_management::update_lock_file(
+                        context,
+                        LockCommand::Upgrade,
+                        build_config.install_dir,
+                        build_config.lock_file,
+                        response,
+                    )
+                    .await
+                    {
+                        warn!("Updating lock file for package: {e}")
+                    };
+                };
```

Comment:
Main logic to write to the lock file after `Upgrade` (resp `Publish` below). I want to call attention to my decisions around `warn!`:

- An error coming from `update_lock_file` shouldn't hard ever in this `Publish` command. All may be well in terms of publishing a package to chain, so we shouldn't hard error on failure to write to `Move.lock` or a related side effect.

- But: it's a bad idea to silently fail. I've opted for a `warn!` here, but that's actually a tracing routine. Curious to hear your thoughts on a good DevX strategy here (is `eprintln!` more appropriate here?). CC @stefan-mysten for appropriate ways to achieve "show a warning error, but don't hard error, in terminal after user runs CLI command, if needed".

Reply:
I don't think we want to use tracing for this -- that should be reserved for internal telemetry, I think, but the general principle of print a warning, but don't hard error seems reasonable. I think `eprintln!` seems fine for now? That's what we use for the `UPDATING GIT DEPENDENCY` messages, etc (with some colouring support).

Longer term, we could also consider adopting `miette` in more places (this is the error system that we use for the `sui client ptb` command).

As with the errors I pointed out earlier, it would be good to share some pointers for folks in this situation about how they can get themselves out of it (in this case it would be find out the addresses themselves and run the manual update command).

Reply:
Thanks, this should be better now, please take another look :-)